libyul/YulName.cpp

```diff
@@ -0,0 +1,430 @@
+	{
+		if (builtin->name.str().substr(0, std::string_view("verbatim").size()) == "verbatim")
+		{
+			auto const key = std::make_tuple(builtin->parameters.size(), builtin->returns.size());
+			auto [it, emplaced] = m_verbatimNames.try_emplace(key);
+			if (emplaced)
+			{
+				it->second = deriveName(predefined().verbatim);
+				m_builtinFunctions[it->second] = convertBuiltinFunction(it->second, *builtin);
+			}
+			return it->second;
+		}
+		else
+		{
+			auto const builtinName = nameOfBuiltin(_label);
+			if (builtinName == emptyName())
+			{
+				m_definedLabels.emplace_back(_label);
+				m_names.emplace_back(m_definedLabels.size() - 1, YulNameState::DEFINED);
+				return m_index++;
+			}
+			else
+				return builtinName;
+		}
+	}
+	else
+	{
+		if (auto const name = nameOfLabel(_label); name != emptyName())
+			return name;
+
+		m_definedLabels.emplace_back(_label);
+		m_names.emplace_back(m_definedLabels.size() - 1, YulNameState::DEFINED);
+		return m_index++;
+	}
+}
+
+YulName YulNameRepository::deriveName(YulName const _name)
+{
+	yulAssert(nameWithinBounds(_name), "YulName exceeds repository size, probably stems from another instance.");
+	m_names.emplace_back(_name, YulNameState::DERIVED);
+	return m_index++;
+}
+
+YulName YulNameRepository::addGhost()
+{
+	return defineName(fmt::format(FMT_COMPILE("GHOST[{}]"), m_nGhosts++));
+}
+
+bool YulNameRepository::isType(YulName const _name) const {
+	return _name >= m_indexBoundaries.beginTypes && _name < m_indexBoundaries.endTypes;
+}
+
+size_t YulNameRepository::nTypes() const
+{
+	return m_indexBoundaries.endTypes - m_indexBoundaries.beginTypes;
+}
+
+bool YulNameRepository::isEvmDialect() const { return dynamic_cast<EVMDialect const*>(&m_dialect.get()) != nullptr; }
+
+void YulNameRepository::generateLabels(std::set<YulName> const& _usedNames, std::set<std::string> const& _illegal)
+{
+	std::set<std::string> used (m_definedLabels.begin(), m_definedLabels.begin() + static_cast<std::ptrdiff_t>(m_indexBoundaries.endBuiltins));
+	std::set<YulName> toDerive;
+	for (auto const name: _usedNames)
+		if (!isDerivedName(name) || isVerbatimFunction(name))
+		{
+			auto const label = labelOf(name);
+			yulAssert(label.has_value());
+			auto const [_, emplaced] = used.emplace(*label);
+			if (!emplaced)
+			{
+				// there's been a clash (eg by calling generate labels twice), let's remove this name and derive
+				// it instead
+				toDerive.insert(name);
+			}
+		}
+
+	std::vector<std::tuple<std::string, YulName>> generated;
+	auto namesIt = _usedNames.begin();
+	for (size_t name = m_indexBoundaries.endBuiltins; name < m_names.size(); ++name)
+	{
+		if (namesIt != _usedNames.end() && name == *namesIt)
+		{
+			if ((isDerivedName(name) && !isVerbatimFunction(name)) || toDerive.find(name) != toDerive.end())
+			{
+				std::string const baseLabel(baseLabelOf(name));
+				std::string label (baseLabel);
+				size_t bump = 1;
+				while (used.count(label) > 0 || _illegal.count(label) > 0)
```

Comment:
I don't see any usages of `generateLabels()` with the `_illegal` parameter explicitly set covered in tests. Also, we should probably assert somewhere that none of the `_illegal` names are used/generated.

Reply:
true, i actually also missed that already existing labels may be invalid

Reply:
i have added asserts although they are a bit expensive because i have to check all names regarding legality :)